fluent-syntax/src/parser/mod.rs

```diff
@@ -1,14 +1,842 @@
+            value: get_number_literal(ps)?,
+        }
+    } else {
+        ast::VariantKey::Identifier {
+            name: get_identifier(ps)?.name,
+        }
+    };
+
+    ps.skip_blank();
+
+    ps.expect_byte(b']')?;
+
+    Ok(key)
+}
+
+fn get_variants<'p>(ps: &mut ParserStream<'p>) -> Result<Vec<ast::Variant<'p>>> {
+    let mut variants = vec![];
+    let mut has_default = false;
+
+    while ps.is_current_byte(b'*') || ps.is_current_byte(b'[') {
+        let default = ps.take_if(b'*');
+
+        if default {
+            if has_default {
+                return error!(ErrorKind::MultipleDefaultVariants, ps.ptr);
+            } else {
+                has_default = true;
+            }
+        }
+
+        let key = get_variant_key(ps)?;
+
+        let value = get_pattern(ps)?;
+
+        if let Some(value) = value {
+            variants.push(ast::Variant {
+                key,
+                value,
+                default,
+            });
+            ps.skip_blank();
+        } else {
+            return error!(ErrorKind::MissingValue, ps.ptr);
+        }
+    }
+
+    if !has_default {
+        error!(ErrorKind::MissingDefaultVariant, ps.ptr)
+    } else {
+        Ok(variants)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum TextElementTermination {
+    LineFeed,
+    CarriageReturn,
+    PlaceableStart,
+    EOF,
+}
+
+#[derive(Debug)]
+enum PatternElementPointers<'a> {
+    Placeable(ast::Expression<'a>),
+    TextElement(usize, usize, usize, TextElementPosition),
+}
+
+#[derive(Debug, PartialEq)]
+enum TextElementPosition {
+    InitialLineStart,
+    LineStart,
+    Continuation,
+}
+
+#[derive(Debug, PartialEq)]
+enum TextElementType {
+    Blank,
+    NonBlank,
+}
+
+fn get_pattern<'p>(ps: &mut ParserStream<'p>) -> Result<Option<ast::Pattern<'p>>> {
+    let mut elements = vec![];
+    let mut last_non_blank = None;
+    let mut common_indent = None;
+
+    ps.skip_blank_inline();
+
+    let mut text_element_role = if ps.skip_eol() {
+        ps.skip_blank_block();
+        TextElementPosition::LineStart
+    } else {
+        TextElementPosition::InitialLineStart
+    };
+
+    while ps.ptr < ps.length {
+        if ps.is_current_byte(b'{') {
+            if text_element_role == TextElementPosition::LineStart {
+                common_indent = Some(0);
+            }
+            let exp = get_placeable(ps)?;
+            last_non_blank = Some(elements.len());
+            elements.push(PatternElementPointers::Placeable(exp));
+            text_element_role = TextElementPosition::Continuation;
+        } else {
+            let slice_start = ps.ptr;
+            let mut indent = 0;
+            if text_element_role == TextElementPosition::LineStart {
+                indent = ps.skip_blank_inline();
+                if ps.ptr >= ps.length {
+                    break;
+                }
+                let b = ps.source[ps.ptr];
+                if indent == 0 {
+                    if b != b'\n' {
```

Comment:
Nit: Do you think that using a different variable name than `b` here would be a good idea? I keep confusing it with the byte literal syntax, `b'x'`.

Reply:
(You can also use a single `if` here.)